d3a_interface/data_classes.py

```diff
@@ -0,0 +1,442 @@
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+import datetime
+import json
+
+from copy import deepcopy
+from dataclasses import dataclass, field, asdict
+from typing import List, Dict, Optional, Tuple, Union
+from pendulum import DateTime, parse
+
+from d3a_interface.utils import datetime_to_string_incl_seconds, key_in_dict_and_not_none
+
+
+def my_converter(date_obj: DateTime):
+    """Define how to convert datetime objects while serializing to json."""
+    if isinstance(date_obj, DateTime):
+        return date_obj.isoformat()
+    return None
+
+
+class BaseBidOffer:
+    """Base class defining shared functionality of Bid and Offer market structures."""
+    def __init__(self, id: str, time: datetime, price: float, energy: float,
+                 original_price: Optional[float] = None,
+                 attributes: Dict = None, requirements: List[Dict] = None):
+        self.id = str(id)
+        self.time = time
+        self.original_price = original_price or price
+        self.price = price
+        self.energy = energy
+        self.energy_rate = self.price / self.energy
+        self.attributes = attributes
+        self.requirements = requirements
+
+    def update_price(self, price: float) -> None:
+        self.price = price
+        self.energy_rate = self.price / self.energy
+
+    def update_energy(self, energy: float) -> None:
+        self.energy = energy
+        self.energy_rate = self.price / self.energy
+
+    def to_json_string(self, **kwargs) -> str:
+        """Convert the Offer or Bid object into its JSON representation.
+
+        Args:
+            **kwargs: additional key-value pairs to be added to the JSON representation.
+        """
+        obj_dict = deepcopy(self.__dict__)
+        if kwargs:
+            obj_dict = {**obj_dict, **kwargs}
+
+        obj_dict["type"] = self.__class__.__name__
+
+        return json.dumps(obj_dict, default=my_converter)
+
+    def serializable_dict(self) -> Dict:
+        return {
+            "type": self.__class__.__name__,
+            "id": self.id,
+            "energy": self.energy,
+            "energy_rate": self.energy_rate,
+            "original_price": self.original_price,
+            "time": datetime_to_string_incl_seconds(self.time),
+            "attributes": self.attributes,
+            "requirements": self.requirements
+        }
+
+    @classmethod
+    def from_json(cls, offer_or_bid, current_time=None) -> Union["Offer", "Bid"]:
+        offer_bid_dict = json.loads(offer_or_bid)
+        object_type = offer_bid_dict.pop("type")
+
+        offer_bid_dict.pop("energy_rate", None)
+
+        if object_type == "Offer":
+            offer_bid_dict["time"] = current_time
```

Comment:
Why time only needs to be injected for Offer and not for Bid ?

Reply:
Should be for both the same. However keep in mind that this is probably the time in string format, json.loads is not able to deserialize datetimes.